### PR TITLE
session: add some compatibility checks for non-transactional DML

### PR DESCRIPTION
**Pull request by ekexium**
Signed-off-by: ekexium <ekexium@gmail.com>

### What problem does this PR solve?

Issue Number: ref #33485

Problem Summary:

A Non-transactional DML is not a SELECT. `sql_select_limit` can limit the select query that is used to generate shard jobs and can lead to unexpected results for a "DML". So it should be ignored when executing the query.

### What is changed and how it works?

- Ignores `sql_select_limit` and `tidb_read_staleness`.
- Returns an error when `tidb_snapshot`, batch-dml, or weak read consistency is used.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by cfzjywxk** [comment]
Would there be some other session variables which should be skipped too?

File: session/nontransactional.go [hunk 1]
```diff
@@ -270,7 +271,11 @@ func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDeleteStmt, s
 		shardColumnCollate = ""
 	}
 
+	// A NT-DML is not a SELECT. We ignore the SelectLimit for selectSQL so that it can read all values.
+	originalSelectLimit := se.GetSessionVars().SelectLimit
```
Comment (you06) [inline session/nontransactional.go:275]: Can we return an error if the limit field is set?

**Comment by ekexium** [comment]
> Would there be some other session variables which should be skipped too?

Maybe `tidb_read_staleness`. If we consider NT-DML as a atomic write operation, then it should not be affected by the variable that is supposed to only affect SELECTs.

I didn't find other session variables that should be ignored. But there are some that I think are worth consideration:
- `tidb_read_consistency`, I suppose we just don't allow NT-DML if it is set to `weak`. I would be hard to understand its meaning
- The deprecated `enable-batch-dml` configuration. I prefer to forbid NT-DML when batch-dml is used. 
- `tidb_snapshot`. If it's set, an error should be returned like other DML.

File: session/nontransactional.go [hunk 2]
```diff
@@ -69,6 +71,9 @@ func HandleNonTransactionalDelete(ctx context.Context, stmt *ast.NonTransactiona
 		return nil, errors.Errorf("non-transactional statement can only run in auto-commit mode. auto-commit:%v, inTxn:%v",
 			se.GetSessionVars().IsAutocommit(), se.GetSessionVars().InTxn())
 	}
```
Comment (you06) [inline session/nontransactional.go:73]: Move the in-txn check into `checkConstraint`?

**Comment by ekexium** [comment]
PTAL @you06 @cfzjywxk

File: session/nontransactional_test.go [hunk 3]
```diff
@@ -226,3 +228,81 @@ func TestNonTransactionalDeleteInvisibleIndex(t *testing.T) {
...
+	tk.MustExec("create table t(a int, b int, key(a))")
+	for i := 0; i < 100; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values (%d, %d)", i, i*2))
+	}
+	tk.MustExec("split on a limit 10 delete from t")
+	tk.MustExec("set @@tidb_read_staleness=0")
+	tk.MustQuery("select count(*) from t").Check(testkit.Rows("0"))
+}
+
+func TestNonTransactionalDeleteTiDBSnapshotAndReadConsistency(t *testing.T) {
```
Comment (cfzjywxk) [inline session/nontransactional_test.go:263]: The first two error return path in `checkConstraint` is not covered by this case, should we also cover them here and rename this test to `TestCheckConstraint`?

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Review by you06 — APPROVED** [review]

**Pull request merged**